File: algorithms/Javascript/twoNumberSum.js
// using nested loops: O(n^2) time | O(1) space
function twoNumberSum(array, targetSum) {
	for (let i = 0; i < array.length - 1; i++) {
		let firstNum = array[i];
		for (let j = i + 1; j < array.length; j++) {
			let secondNum = array[j];
			if (firstNum + secondNum === targetSum) {
				return [firstNum, secondNum].sort((a, b) => a - b);
			}
		}
	}
	return [];
}

// using a hash: O(n) time | O(n) space
function twoSumNumbers(array, targetSum) {
	let nums = {};
	for (let i = 0; i < array.length; i++) {
		let currentNum = array[i];
		let potentialMatch = targetSum - currentNum;
		if (nums[potentialMatch]) {
			return [potentialMatch, currentNum].sort((a, b) => a - b);
		} else {
			nums[currentNum] = true;
		}
	}
	return [];
}

File: algorithms/Python/twoNumberSum.py
# O(n^2) time | O(1) space
def twoNumberSum(array, targetSum):
    for i in range(len(array) - 1):
        firstNum = array[i]
        for j in range(i + 1, len(array)):
            secondNum = array[j]
            if firstNum + secondNum == targetSum:
                return sorted([firstNum, secondNum])
    return []

# O(n) time | O(n) space
def twoNumberSum2(array, targetSum):
    nums = {}
    for num in array:
        potentialMatch = targetSum - num
        if potentialMatch in nums:
            return sorted([potentialMatch, num])
        else: # store num in hash
            nums[num] = True
    return []
# e.g. twoNumberSum([1, 4, 2, 6, 9], 11)
